src/aio.rs

```diff
@@ -577,3 +577,91 @@ impl ConnectionLike for MultiplexedConnection {
         self.db
     }
 }
+
+#[cfg(feature = "tokio-rt-core")]
+mod connection_manager {
+    use super::*;
+
+    /// A `ConnectionManager` is a proxy that wraps a multiplexed connection and
+    /// automatically reconnects to the server when necessary.
+    pub struct ConnectionManager {
+        connection_info: ConnectionInfo,
+        connection: MultiplexedConnection,
+    }
+
+    #[cfg(feature = "tokio-rt-core")]
+    impl ConnectionManager {
+        /// Connect to the server and store the connection inside the returned `ConnectionManager`.
+        ///
+        /// TODO: Add `new_tokio`?
+        pub async fn new(connection_info: ConnectionInfo) -> RedisResult<Self> {
+            let (connection, driver) = MultiplexedConnection::new(connect(&connection_info).await?);
+            tokio::spawn(driver);
+            Ok(Self {
+                connection_info,
+                connection,
+            })
+        }
+
+        async fn reconnect(&mut self) -> RedisResult<&mut Self> {
+            let (new_connection, driver) = MultiplexedConnection::new(connect(&self.connection_info).await?);
+            tokio::spawn(driver);
+            let _old_connection = std::mem::replace(&mut self.connection, new_connection);
+            Ok(self)
+        }
+    }
+
+    #[cfg(feature = "tokio-rt-core")]
+    impl ConnectionLike for ConnectionManager {
+        fn req_packed_command<'a>(&'a mut self, cmd: &'a Cmd) -> RedisFuture<'a, Value> {
+            (async move {
+                let result = self.connection.req_packed_command(cmd).await;
+                if let Err(ref e) = result {
+                    if e.is_connection_dropped() {
+                        // TODO: This could lead to multiple reconnection
+                        // attempts being done in parallel... We should
+                        // probably avoid that.
+                        println!("Connection lost, reconnecting");
+                        if let Err(reconnect_failed) = self.reconnect().await {
```

Comment:
This should probably happen asynchronously, but I had a lifetime issue when I put the `self.reconnect()` call into `tokio::spawn`.

Reply:
You will need to put the contents of `ConnectionManager` behind an `Arc`. As it is you can only issue one request at a time on the `ConnectionManager`.

That will then let you clone `self` when needed to get a 'static value for `tokio::spawn`.